corehq/apps/userreports/sql/adapter.py

```diff
@@ -48,20 +48,26 @@ def save(self, doc):
         indicator_rows = self.config.get_all_values(doc)
         if indicator_rows:
             table = self.get_table()
-            with self.engine.begin() as connection:
-                # delete all existing rows for this doc to ensure we aren't left with stale data
-                delete = table.delete(table.c.doc_id == doc['_id'])
-                connection.execute(delete)
-                for indicator_row in indicator_rows:
-                    all_values = {i.column.database_column_name: i.value for i in indicator_row}
-                    insert = table.insert().values(**all_values)
-                    try:
-                        connection.execute(insert)
-                    except IntegrityError:
-                        # Someone beat us to it. Concurrent inserts can happen
-                        # when a doc is processed by the celery rebuild task
-                        # at the same time as the pillow.
-                        pass
+            try:
+                with self.engine.begin() as connection:
+                    # delete all existing rows for this doc to ensure we aren't left with stale data
+                    delete = table.delete(table.c.doc_id == doc['_id'])
+                    connection.execute(delete)
+                    bad_saves = False
+                    for indicator_row in indicator_rows:
+                        all_values = {i.column.database_column_name: i.value for i in indicator_row}
+                        insert = table.insert().values(**all_values)
+                        try:
+                            connection.execute(insert)
+                        except (IntegrityError, InternalError):
+                            bad_saves = True
+                    if bad_saves:
+                        raise BadSaveError()
```

Comment:
Thanks @millerdev. So I think there are some tricky requirements that would be good to iron out before we get into how the code should be constructed.

> It is troubling that you can call indicator_thing.save(doc) and it will return as if everything is fine when in fact the save failed.

I don't disagree with this, but I wanted to make sure you understood the surrounding code and situation well enough to understand how we could better improve this.

1. This code is called on a user-configured data source on an arbitrary document that is supposed to match that data source.
2. As such, it can easily fail if, e.g. the user has configured a column to expect a number and it finds a string
3. The data source can also be configured to save multiple rows at a time (in the for loop) where those rows again can be determined dynamically based on the properties of the data source and the document going in.
4. We have received feedback from people using the framework at various points that they prefer a "best effort" approach to saving the data, and would prefer even partial saves as a result of (3) to get in.

In an ideal world we have a whole framework that can detect these user-created issues and help the user find and resolve them. However, we don't have that today and I believe it will be a large effort to build it out. In the meantime we either need to build a solution that fits these requirements or change the requirements. For the latter, dropping requirement 4 definitely makes the code quite a bit cleaner, and allows us to do what you're proposing. However I don't know how to evaluate the impact of that change on our users against the impact on us as code maintainers.

Do you have any suggestions on how you think we should be thinking about that? Or alternatively do you think you can come up with a cleaner approach that includes requirement 4?